### PR TITLE
issue751: store g-values in GEvaluator

### PR DESCRIPTION
**Pull request by jendrikseipp**
For now, I only tackled eager search.

### CODE REVIEW

File: src/search/search_engines/plugin_astar.cc [hunk 1]
```diff
@@ -38,9 +42,14 @@ static shared_ptr<SearchEngine> _parse(OptionParser &parser) {
 
     shared_ptr<eager_search::EagerSearch> engine;
     if (!parser.dry_run()) {
-        auto temp = search_common::create_astar_open_list_factory_and_f_eval(opts);
-        opts.set("open", temp.first);
-        opts.set("f_eval", temp.second);
+        shared_ptr<Evaluator> g = make_shared<g_evaluator::GEvaluator>(
```
Comment (FlorianPommerening) [inline src/search/search_engines/plugin_astar.cc:45]: This change seems unrelated to the issue. Pulling out the computation of "f = g + h" defeats the purpose of a function that specifically creates the open list of A*, in my opinion, i.e., that part is what makes it an A* open list..

File: src/search/search_engines/plugin_astar.cc [hunk 2]
```diff
@@ -38,9 +42,14 @@ static shared_ptr<SearchEngine> _parse(OptionParser &parser) {
...
+            tasks::g_root_task);
```
Comment (FlorianPommerening) [inline src/search/search_engines/plugin_astar.cc:46]: Is it necessary to use the root task here?

File: src/search/search_engines/search_common.cc [hunk 3]
```diff
@@ -98,7 +99,7 @@ shared_ptr<OpenListFactory> create_wastar_open_list_factory(
         options.get_list<shared_ptr<Evaluator>>("evals");
     int w = options.get<int>("w");
 
-    shared_ptr<GEval> g_eval = make_shared<GEval>();
+    shared_ptr<GEval> g_eval = make_shared<GEval>(tasks::g_root_task);
```
Comment (FlorianPommerening) [inline src/search/search_engines/search_common.cc:102]: see above

File: src/search/search_engines/search_common.h [hunk 4]
```diff
@@ -68,15 +68,13 @@ extern std::shared_ptr<OpenListFactory> create_wastar_open_list_factory(
...
 
 /*
-  Create open list factory and f_evaluator (used for displaying progress
-  statistics) for A* search.
+  Create open list factory for A* search.
 
   The resulting open list factory produces a tie-breaking open list
-  ordered primarily on g + h and secondarily on h. Uses "eval" from
-  the passed-in Options object as the h evaluator.
+  ordered primarily on g + h and secondarily on h.
```
Comment (FlorianPommerening) [inline src/search/search_engines/search_common.h:74]: This is the part where I don't like the way the new function is split up: the comment says that the open list is ordered by g+ h, but the parameters of the functions are h and f, so the comment is only true if the user puts in the correct f.

File: src/search/search_node_info.cc [hunk 5]
```diff
@@ -1,10 +1,6 @@
...
 
-static const int pointer_bytes = sizeof(void *);
-static const int info_bytes = 3 * sizeof(int) + sizeof(StateID);
-static const int padding_bytes = info_bytes % pointer_bytes;
-
 static_assert(
-    sizeof(SearchNodeInfo) == info_bytes + padding_bytes,
-    "The size of SearchNodeInfo is larger than expected. This probably means "
-    "that packing two fields into one integer using bitfields is not supported.");
+    sizeof(SearchNodeInfo) == 4 + sizeof(StateID) + sizeof(OperatorID),
```
Comment (FlorianPommerening) [inline src/search/search_node_info.cc:4]: This ignores word alignment now.

File: src/search/search_node_info.h [hunk 6]
```diff
@@ -10,15 +10,14 @@
 struct SearchNodeInfo {
     enum NodeStatus {NEW = 0, OPEN = 1, CLOSED = 2, DEAD_END = 3};
 
-    unsigned int status : 2;
-    int g : 30;
+    // TODO: pack this into 8 bytes.
```
Comment (FlorianPommerening) [inline src/search/search_node_info.h:13]: This might be more tricky than expected with the custom ID classes.

File: src/search/search_node_info.h [hunk 7]
```diff
@@ -10,15 +10,14 @@
...
+    NodeStatus status;
     StateID parent_state_id;
     OperatorID creating_operator;
-    int real_g;
 
     SearchNodeInfo()
-        : status(NEW), g(-1), parent_state_id(StateID::no_state),
-          creating_operator(-1), real_g(-1) {
+        : status(NEW), parent_state_id(StateID::no_state),
+          creating_operator(OperatorID::no_operator) {
```
Comment (FlorianPommerening) [inline src/search/search_node_info.h:20]: Nice, we should have done this when we introduced the class.

File: src/search/evaluators/g_evaluator.cc [hunk 8]
```diff
@@ -5,15 +5,40 @@
...
 
+#include "../tasks/root_task.h"
+
 using namespace std;
 
 namespace g_evaluator {
+static const int INFTY = numeric_limits<int>::max();
+
+GEvaluator::GEvaluator(const shared_ptr<AbstractTask> &task)
+    : task(task),
```
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/evaluators/g_evaluator.cc:16]: I agree. Do you have a solution for this?

File: src/search/search_engine.cc [hunk 9]
```diff
@@ -60,6 +60,9 @@ SearchEngine::SearchEngine(const Options &opts)
         utils::exit_with(ExitCode::SEARCH_INPUT_ERROR);
     }
     bound = opts.get<int>("bound");
+    if (bound != numeric_limits<int>::max()) {
+        real_g_evaluator = make_shared<g_evaluator::GEvaluator>(task);
```
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/search_engine.cc:64]: Good point! I'll add a special case for that.

File: src/search/search_engines/eager_search.cc [hunk 10]
```diff
@@ -182,28 +200,34 @@ SearchStatus EagerSearch::step() {
...
         bool is_preferred = preferred_operators.contains(op_id);
 
-        SearchNode succ_node = search_space.get_node(succ_state);
-
+        int succ_g_old = EvaluationContext(succ_state).get_evaluator_value_or_infinity(
+            g_evaluator.get());
         for (Evaluator *evaluator : path_dependent_evaluators) {
             evaluator->notify_state_transition(s, op_id, succ_state);
         }
+        EvaluationContext succ_eval_context(succ_state, is_preferred, &statistics);
```
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/search_engines/eager_search.cc:227]: I also thought so. But I didn't see a clean workaround. Do you?

File: src/search/search_engines/plugin_astar.cc (hunk 1 above)
Comment (jendrikseipp) [reply src/search/search_engines/plugin_astar.cc:45]: I split it because I didn't see a good way to return the open list, the f-evaluator and the g-evaluator from the function. So you see a good way to do so? Or should I pass the g-evaluator to the function (removing the need to return it)?

File: src/search/search_engines/plugin_astar.cc (hunk 2 above)
Comment (jendrikseipp) [reply src/search/search_engines/plugin_astar.cc:46]: We can't use the task from the search engine, because it is hard-coded (to g_root_task) in the search engine ctor, and not passed to the search engine via the options.

File: src/search/search_node_info.cc (hunk 5 above)
Comment (jendrikseipp) [reply src/search/search_node_info.cc:4]: I thought so as well, but this version of the code compiles. My guess is that there will be padding once more than one SearchNodeInfo is concatenated in memory.

File: src/search/search_engines/eager_search.cc [hunk 11]
```diff
@@ -33,6 +38,15 @@ EagerSearch::EagerSearch(const Options &opts)
...
         utils::exit_with(utils::ExitCode::SEARCH_INPUT_ERROR);
     }
+    // Even for GBFS, we need a g-evaluator for reporting progress.
+    if (!g_evaluator) {
+        shared_ptr<AbstractTask> task = tasks::g_root_task;
+        OperatorCost cost_type = opts.get<OperatorCost>("cost_type");
+        if (cost_type != OperatorCost::NORMAL) {
+            task = make_shared<tasks::CostAdaptedTask>(task, cost_type);
+        }
+        g_evaluator = make_shared<g_evaluator::GEvaluator>(task);
```
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/search_engines/eager_search.cc:48]: I agree, this is definitely problematic in the current patch. I see two options: 1) we remove cost_type from the options and expose g_eval on the commandline (breaking backwards compatibility of the configs). 2) we make sure that users can never pass cost_type and g_eval at the same time. I tried to use 2) in the patch, but I forgot to use the cost_type for creating the g_eval. I'll fix that. I'd like to avoid 1) to avoid breaking old configs, but maybe we'll see later that we have to break them.

File: src/search/search_engines/plugin_astar.cc (hunk 1 above)
Comment (jendrikseipp) [reply src/search/search_engines/plugin_astar.cc:45]: I'm working on a different solution that restores the original function and directly inserts the g evaluators into the options object.

File: src/search/evaluators/g_evaluator.cc [hunk 12]
```diff
@@ -22,7 +47,7 @@ static shared_ptr<Evaluator> _parse(OptionParser &parser) {
     if (parser.dry_run())
         return nullptr;
     else
-        return make_shared<GEvaluator>();
+        return make_shared<GEvaluator>(tasks::g_root_task);
```
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/evaluators/g_evaluator.cc:50]: This code is only relevant if users use g() on the commandline. It is unrelated to the cost_type option of the search engines. It can still lead to confusion though, I agree.

File: src/search/evaluators/g_evaluator.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/evaluators/g_evaluator.cc:16]: We decided to make GEvaluator a subclass of Heuristic (which will be renamed to CachingEvaluator), which should solved this problem.

File: src/search/search_engines/eager_search.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/search_engines/eager_search.cc]: We decided to switch to option 1).

File: src/search/search_engines/eager_search.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (jendrikseipp) [reply src/search/search_engines/eager_search.cc:222]: This will be resolved once GEvaluator caches its estimates and knows which ones are "dirty".

**Comment by ClemensBuechner** [comment]
We decided to close this pull request because it is outdated (see issue tracker: https://issues.fast-downward.org/issue751).
